src/main/kotlin/com/grepp/nbe1_3_team04/chat/service/ChatroomServiceImpl.kt

```diff
@@ -0,0 +1,115 @@
+import com.grepp.nbe1_3_team04.chat.service.response.ChatroomResponse
+import com.grepp.nbe1_3_team04.global.exception.ExceptionMessage
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+
+
+@Service
+class ChatroomServiceImpl(
+    private val chatroomRepository: ChatroomRepository,
+    private val redisChatroomRepository: RedisChatroomRepository
+) : ChatroomService {
+
+    /**
+     * 채팅방 생성
+     * 채팅방을 만든 사람은 자동으로 채팅방 초대
+     *
+     * @param request 채팅방 이름 ( 팀 or 예약 만들어질 때 이름 넣어서 보내주기 )
+     * @return
+     */
+    @Transactional
+    override fun createChatroom(request: ChatroomServiceRequest): ChatroomResponse {
+        val chatroom: Chatroom = chatroomRepository.save(Chatroom.create(request.name))
+
+        // redis Hash에 저장
+        redisChatroomRepository.createChatRoom(chatroom)
+
+        return ChatroomResponse(chatroom)
+    }
+
+    @Transactional
+    override fun createTeamChatroom(request: ChatroomServiceRequest, teamId: Long): ChatroomResponse {
+        val chatroom: Chatroom = chatroomRepository.save(TeamChatroom.create(request.name, teamId))
+
+        // redis Hash에 저장
+        redisChatroomRepository.createChatRoom(chatroom)
+
+        return ChatroomResponse(chatroom)
+    }
+
+    @Transactional
+    override fun createReservationChatroom(request: ChatroomServiceRequest, reservationId: Long): ChatroomResponse {
+        val chatroom: Chatroom = chatroomRepository.save(ReservationChatroom.create(request.name, reservationId))
+
+        // redis Hash에 저장
+        redisChatroomRepository.createChatRoom(chatroom)
+
+        return ChatroomResponse(chatroom)
+    }
+
+    /**
+     * 채팅방 삭제 ( 채팅방 인원 삭제도 같이 진행해야 함 )
+     *
+     * @param chatroomId
+     * @return
+     */
+    @Transactional
+    override fun deleteChatroomByChatroomId(chatroomId: Long): Long {
+        val chatroom = getChatroomByChatroomId(chatroomId)
+
+        return deleteChatroom(chatroom)
+    }
+
+    @Transactional
+    override fun deleteTeamChatroom(teamId: Long): Long {
+        val chatroom = getChatroomByTeamId(teamId)
+
+        return deleteChatroom(chatroom)
+    }
+
+    @Transactional
+    override fun deleteReservationChatroom(reservationId: Long): Long {
+        val chatroom = getChatroomByReservationId(reservationId)
+
+        return deleteChatroom(chatroom)
+    }
+
+    /**
+     * 채팅방 수정
+     */
+    @Transactional
+    override fun updateChatroom(chatroomId: Long, request: ChatroomServiceRequest): ChatroomResponse {
+        val chatroom = getChatroomByChatroomId(chatroomId)
+
+        chatroom.updateName(request.name)
+
+        return ChatroomResponse(chatroom)
+    }
+
+    private fun deleteChatroom(chatroom: Chatroom): Long {
+        redisChatroomRepository.deleteChatroomFromRedis(chatroom.chatroomId!!)
+
+        chatroomRepository.deleteById(chatroom.chatroomId!!)
+        return chatroom.chatroomId!!
```

Comment:
여기 부분 !! 는 앞에 붙였는데도 또 붙여줘야 하나요??

Reply:
넵 chatroom 엔티티 클래스의 chatroomId는 nullable 하기에...

Reply:
그렇군요.! 앞에서 !! 붙여주면 뒤에서 안해줘도 되는 것들도 있길래 이것도 동일한 줄 알았습니다

Reply:
> 그렇군요.! 앞에서 !! 붙여주면 뒤에서 안해줘도 되는 것들도 있길래 이것도 동일한 줄 알았습니다

이거 뒤에 붙는건 아마도 변수로 선언해서 하셨으면 반복해서 안해도 괜찮으셨을건데, 사용할 때 마다 꺼내서 null 아님을 표시해주고 있어서 이렇게 해주고있는것 같아요

```suggestion
        val chatroomId = chatroom.chatroomId!!
        chatroomRepository.deleteById(chatroomId)
        return chatroomId
```

이렇게 하셨으면 한번만 선언해도 될거같다는 생각이들긴하네요